returnn/torch/frontend/frontend.py

```diff
@@ -17,3 +24,74 @@ class TorchFrontend(Frontend[torch.Tensor]):
     """
 
     RawTensorType = torch.Tensor
+    _internal_frontend = TorchInternalFrontend
+
+    @staticmethod
+    def dot(a: _TT, b: _TT, *, reduce: Union[Dim, Sequence[Dim]]) -> _TT:
+        """
+        dot-product of a and b, see base class doc string
+        """
+        if isinstance(reduce, Dim):
+            reduce = [reduce]
+
+        a_dims = a.dims
+        b_dims = b.dims
+
+        assert all(
+            dim in a_dims for dim in reduce
+        ), f"'a' does not have the specified reduce dim(s) {reduce} (a dims: {a_dims})"
+        assert all(
+            dim in b_dims for dim in reduce
+        ), f"'b' does not have the specified reduce dim(s) {reduce} (b dims: {b_dims})"
+
+        reduce_dims_sorted_by_a = [dim for dim in a_dims if dim in reduce]
+
+        a_reduce_axes = [a_dims.index(reduce_dim) for reduce_dim in reduce_dims_sorted_by_a]
+        b_reduce_axes = [b_dims.index(reduce_dim) for reduce_dim in reduce_dims_sorted_by_a]
+
+        a_unique_axes = [i for (i, dim) in enumerate(a_dims) if dim not in b_dims]
+        b_unique_axes = [i for (i, dim) in enumerate(b_dims) if dim not in a_dims]
+
+        common_dims = [dim for dim in a_dims if dim in b_dims and dim not in reduce]
+
+        a_common_axes = [a_dims.index(common_dim) for common_dim in common_dims]
+        b_common_axes = [b_dims.index(common_dim) for common_dim in common_dims]
+
+        a_raw = a.raw_tensor
+        b_raw = b.raw_tensor
+
+        a_shape = a_raw.shape
+        b_shape = b_raw.shape
+
+        common_axes_shape = tuple(a_shape[i] for i in a_common_axes)
+        b_common_axes_shape = tuple(b_shape[i] for i in b_common_axes)
+        assert common_axes_shape == b_common_axes_shape, "Tensor shape for common Dims of a and b does not match."
+
+        a_unique_axes_shape = tuple(a_shape[i] for i in a_unique_axes)
+        b_unique_axes_shape = tuple(b_shape[i] for i in b_unique_axes)
+
+        a_unique_axes_total_dimension = prod(a_unique_axes_shape)
+        b_unique_axes_total_dimension = prod(b_unique_axes_shape)
+
+        reduce_axes_shape = tuple(a_shape[i] for i in a_reduce_axes)
+        b_reduce_axes_shape = tuple(b_shape[i] for i in b_reduce_axes)
+        assert reduce_axes_shape == b_reduce_axes_shape, "Tensor shape for reduce Dims does not match between a and b."
+
+        reduce_axes_total_dimension = prod(reduce_axes_shape)
+
+        a_raw = torch.permute(a_raw, a_common_axes + a_unique_axes + a_reduce_axes)
+        a_raw = torch.reshape(a_raw, common_axes_shape + (a_unique_axes_total_dimension, reduce_axes_total_dimension))
+
+        b_raw = torch.permute(b_raw, b_common_axes + b_reduce_axes + b_unique_axes)
+        b_raw = torch.reshape(b_raw, common_axes_shape + (reduce_axes_total_dimension, b_unique_axes_total_dimension))
+
+        raw_result = torch.matmul(a_raw, b_raw)
```

Comment:
I'm not sure if we really want to call `matmul` here. I think calling either `torch.bmm` or `torch.mm` directly makes the code more clear, maybe faster, also more bug-proof.

For `torch.bmm`, I think you need to merge all the `common_axes_shape` together. If that product is 1, you can just call `torch.mm`.

Reply:
Hmm, if there's a difference at all I would assume it is slower to do the shape check and the extra reshape from Python. It's true that matmul has many different behaviours, but the special cases are about 1-D inputs and broadcasting. Here we have neither. I think this is the implementation: https://github.com/pytorch/pytorch/blob/758ad516f32708de243f194144ee0f7b9e0f5117/aten/src/ATen/native/LinearAlgebra.cpp#L340
There is a direct fallback to `mm` and reshaping before calling `bmm`, so pretty much exactly what we would do here. But it's also obvious how to do it here, so I can change it if you want.

Reply:
There is quite some logic involved in matmul, and we don't need any of that, we know directly how to call `bmm` or `mm`. Also, at least for me, when looking at the matmul code or documentation, it is not immediately obvious that it really does what we need.

Reply:
Ok, changed it.